### PR TITLE
Add timeout option

### LINKED ISSUE #129: Timeout for getting files from IPFS
Hi,

Is there any way to specify timeout for methods like cat()/get() or any other way to handle calls to this functions with hash to non-existing resource?

Thanks

### PR DESCRIPTION
**Pull request by AliabbasMerchant**
Fixes #129   
@alexander255 Please close #143   

Also, most probably a problem with my system, but _all_ the tests are failing. I am getting a `UnicodeDecodeError`.  
Please check once before merging.

### CODE REVIEW

File: test/unit/test_http.py [hunk 1]
```diff
@@ -222,6 +240,18 @@ def test_failed_download(http_client):
 		with pytest.raises(ipfshttpclient.exceptions.StatusError):
 			http_client.download("/fail")
 
+def test_download_timeout(self):
```
Comment (alexander255) [inline test/unit/test_http.py]: Drop the `self` here to make the testing work (this is not inside a class, so there is no `self` to refer to here)

File: test/unit/test_http.py [hunk 2]
```diff
@@ -222,6 +240,18 @@ def test_failed_download(http_client):
...
+        """Tests that a timed-out download raises a TimeoutError."""
+        with HTTMock(return_timeout_2_sec):
+                with pytest.raises(ipfshttpclient.exceptions.TimeoutError):
+                        self.client.download('/timeout', timeout=1)
+
+def test_request_timeout(self):
```
Comment (alexander255) [inline test/unit/test_http.py]: Same here

File: test/unit/test_http.py (hunk 1 above)
Comment (AliabbasMerchant) [reply test/unit/test_http.py]: Oops.. XD

**Comment by AliabbasMerchant** [comment]
@alexander255 Done

My stack trace:
```
Launching pytest with arguments test_http.py::test_stream in E:\projects_workspaces\IPFS\py-ipfs-api\test\unit

Traceback (most recent call last):
  File "C:\Program Files\JetBrains\PyCharm Community Edition 2018.3.2\helpers\pycharm\_jb_pytest_runner.py", line 31, in <module>
    pytest.main(args, plugins_to_load)
  File "C:\Users\aliab\AppData\Local\Programs\Python\Python37-32\lib\site-packages\_pytest\config\__init__.py", line 58, in main
    config = _prepareconfig(args, plugins)
  File "C:\Users\aliab\AppData\Local\Programs\Python\Python37-32\lib\site-packages\_pytest\config\__init__.py", line 196, in _prepareconfig
    pluginmanager=pluginmanager, args=args
  File "C:\Users\aliab\AppData\Local\Programs\Python\Python37-32\lib\site-packages\pluggy\hooks.py", line 284, in __call__
    return self._hookexec(self, self.get_hookimpls(), kwargs)
  File "C:\Users\aliab\AppData\Local\Programs\Python\Python37-32\lib\site-packages\pluggy\manager.py", line 67, in _hookexec
    return self._inner_hookexec(hook, methods, kwargs)
  File "C:\Users\aliab\AppData\Local\Programs\Python\Python37-32\lib\site-packages\pluggy\manager.py", line 61, in <lambda>
    firstresult=hook.spec.opts.get("firstresult") if hook.spec else False,
  File "C:\Users\aliab\AppData\Local\Programs\Python\Python37-32\lib\site-packages\pluggy\callers.py", line 203, in _multicall
    gen.send(outcome)
  File "C:\Users\aliab\AppData\Local\Programs\Python\Python37-32\lib\site-packages\_pytest\helpconfig.py", line 93, in pytest_cmdline_parse
    config = outcome.get_result()
  File "C:\Users\aliab\AppData\Local\Programs\Python\Python37-32\lib\site-packages\pluggy\callers.py", line 80, in get_result
    raise ex[1].with_traceback(ex[2])
  File "C:\Users\aliab\AppData\Local\Programs\Python\Python37-32\lib\site-packages\pluggy\callers.py", line 187, in _multicall
    res = hook_impl.function(*args)
  File "C:\Users\aliab\AppData\Local\Programs\Python\Python37-32\lib\site-packages\_pytest\config\__init__.py", line 675, in pytest_cmdline_parse
    self.parse(args)
  File "C:\Users\aliab\AppData\Local\Programs\Python\Python37-32\lib\site-packages\_pytest\config\__init__.py", line 845, in parse
    self._preparse(args, addopts=addopts)
  File "C:\Users\aliab\AppData\Local\Programs\Python\Python37-32\lib\site-packages\_pytest\config\__init__.py", line 790, in _preparse
    self._initini(args)
  File "C:\Users\aliab\AppData\Local\Programs\Python\Python37-32\lib\site-packages\_pytest\config\__init__.py", line 729, in _initini
    config=self,
  File "C:\Users\aliab\AppData\Local\Programs\Python\Python37-32\lib\site-packages\_pytest\config\findpaths.py", line 127, in determine_setup
    rootdir, inifile, inicfg = getcfg([ancestor], config=config)
  File "C:\Users\aliab\AppData\Local\Programs\Python\Python37-32\lib\site-packages\_pytest\config\findpaths.py", line 34, in getcfg
    iniconfig = py.iniconfig.IniConfig(p)
  File "C:\Users\aliab\AppData\Local\Programs\Python\Python37-32\lib\site-packages\py\_vendored_packages\iniconfig.py", line 54, in __init__
    tokens = self._parse(iter(f))
  File "C:\Users\aliab\AppData\Local\Programs\Python\Python37-32\lib\site-packages\py\_vendored_packages\iniconfig.py", line 82, in _parse
    for lineno, line in enumerate(line_iter):
  File "C:\Users\aliab\AppData\Local\Programs\Python\Python37-32\lib\encodings\cp1252.py", line 23, in decode
    return codecs.charmap_decode(input,self.errors,decoding_table)[0]
UnicodeDecodeError: 'charmap' codec can't decode byte 0x90 in position 1421: character maps to <undefined>
```
Probably a PyCharm error

**Comment by alexander255** [comment]
@AliabbasMerchant: Thanks for the stack trace! It appears that the default system encoding on your system is `cp1252` (aka `windows-1252`, sometimes also incorrectly called `iso-8859-1`), while the `tox.ini` file shipped with the project uses `utf-8`.

Is the same behaviour exhibited when running the tests with `tox` from the command-line? (Test that by installing [Python for Windows](https://www.python.org/downloads/windows/), then using `pip install tox` to install `tox` and finally running `tox` from within the *py-ipfs-http-client* source directory.)

Could you also post the output of the following code snippet from both PyCharm and a native Python installation:

```py
import locale
print(locale.getpreferredencoding())
```

I'll look into why the tests still fail on CI, since everything appears like it should be working. :+1:

**Comment by alexander255** [comment]
TL;DR why the test isn't working: `HTTMock` does not actually emulate a web server or anything, it simply patches some `requests` methods to return dummy responses based on the mocking function return value if their match pattern is matched. (Socket) timeouts are simply not implemented for these mocked methods.

**Comment by AliabbasMerchant** [comment]
@alexander255   
Yes. Default encoding is `cp1252`
For this project, I have not used any venv in Pycharm. Running directly on the system python 😅 
`tox` is also showing the same `UnicodeDecodeError`

> (Socket) timeouts are simply not implemented for these mocked methods.

The same tests were working in the `master` branch (#143).


Anyways, ok. No problem. Merge it if the tests are passing on your PC.   
I will check about my `UnicodeDecodeError`.

**Comment by alexander255** [comment]
Tests aren't passing on my PC either, but I'll be switching all the `test_http.py` tests from `httmock` to `pytest-localserver` which starts an actual dummy HTTP server for testing. This way issues all “emulation” will be perfect, since the only thing spoofed different from a real server is the actual responses sent, and we get rid of another requests dependency I didn't even know we had. :slightly_smiling_face: 

> I will check about my `UnicodeDecodeError`.

The thing is, I fear it's not just *your* `UnicodeDecodeError`: It may affect everybody trying to hack on this library on Windows and also people using a non-UTF-8 locale on Unix.

**Comment by AliabbasMerchant** [comment]
Yeah. The encoding and the difference in the OS are indeed a great issue.  
Even some other tests are failing on windows.  
For example, the file size tests are also failing.

**Comment by alexander255** [comment]
The changes of this PR where commited as part of #170. Regarding Windows support I opened #178.

Closing this.
